src/probnum/filtsmooth/gaussfiltsmooth/kalman.py

```diff
@@ -24,406 +33,186 @@ class Kalman(BayesFiltSmooth):
+    smooth_step
+        Custom update step. Choose between e.g. classical, and square-root implementation. Default is 'rts_smooth_step_classic`.
     """
 
-    def __init__(self, dynamics_model, measurement_model, initrv):
-        """Check that the initial distribution is Gaussian."""
-        if not issubclass(type(initrv), Normal):
-            raise ValueError(
-                "Gaussian filters/smoothers need initial "
-                "random variables with Normal distribution."
-            )
-        super().__init__(dynamics_model, measurement_model, initrv)
-
-    def filtsmooth(self, dataset, times, _intermediate_step=None):
-        """Apply Gaussian filtering and smoothing to a data set.
+    def __init__(
+        self,
+        dynamics_model,
+        measurement_model,
+        initrv,
+        predict=predict_via_transition,
+        measure=measure_via_transition,
+        update=update_classic,
+        smooth_step=rts_smooth_step_classic,
+    ):
+        self.dynamics_model = dynamics_model
+        self.measurement_model = measurement_model
+        self.initrv = initrv
+        self.predict = lambda *args, **kwargs: predict(dynamics_model, *args, **kwargs)
+        self.measure = lambda *args, **kwargs: measure(
+            measurement_model, *args, **kwargs
+        )
+        self.update = lambda *args, **kwargs: update(measurement_model, *args, **kwargs)
+        self.smooth_step = smooth_step
+        super().__init__(
+            dynamics_model=dynamics_model,
+            measurement_model=measurement_model,
+            initrv=initrv,
+        )
 
-        Parameters
-        ----------
-        dataset : array_like, shape (N, M)
-            Data set that is filtered.
-        times : array_like, shape (N,)
-            Temporal locations of the data points.
-            The zeroth element in times and dataset is the location of the initial random variable.
+    def iterated_filtsmooth(
+        self, dataset, times, stopcrit=None, _intermediate_step=None
+    ):
+        """Compute an iterated smoothing estimate with repeated posterior linearisation.
 
-        Returns
-        -------
-        KalmanPosterior
-            Posterior distribution of the smoothed output
+        If the extended Kalman filter is used, this yields the IEKS. In
+        any case, the result is an approximation to the maximum-a-
+        posteriori estimate.
         """
+
+        if stopcrit is None:
+            stopcrit = StoppingCriterion()
+
+        # Initialise iterated smoother
+        old_posterior = self.filtsmooth(
+            dataset=dataset,
+            times=times,
+            _intermediate_step=_intermediate_step,
+            _previous_posterior=None,
+        )
+        new_posterior = old_posterior
+        new_mean = new_posterior.state_rvs.mean
+        old_mean = np.inf * np.ones(new_mean.shape)
+        while stopcrit.do_not_terminate_yet(
```

Comment:
I am not sure about the name of this function, but out of all the options I tried, this was the least confusing. This function basically computes the "internal norm" (you know what I mean), compares it to 1, and throws and error if the maxit is reached.

Reply:
I changed it to `while not stopcrit.terminate`. Tbf, this is how I started writing it but then I got so confused with the trues, falses, and negations, that I changed things around until they worked and then left it there :D Of course, simplifying the working solution again is a good idea.